File: templates/go/content/capabilities/lifecycle.go
package capabilities

import (
	"go.bytecodealliance.org/cm"
	lifecycle "weather_go/internal/wasmcp/mcp/lifecycle"
	lifecycletypes "weather_go/internal/wasmcp/mcp/lifecycle-types"
	mcptypes "weather_go/internal/wasmcp/mcp/mcp-types"
)

// Initialize handles session initialization.
//
// Technical Note: We return a cm.Result instead of Go's idiomatic (value, error)
// because WebAssembly core modules can only return a single value. The Result type
// is defined in WIT as result<T, E> and represents either success (T) or error (E).
// The InitializeResultShape is wit-bindgen-go's internal storage type for the result.
func Initialize(request lifecycletypes.InitializeRequest) cm.Result[lifecycle.InitializeResultShape, lifecycletypes.InitializeResult, mcptypes.McpError] {
	result := lifecycletypes.InitializeResult{
		ProtocolVersion: lifecycletypes.ProtocolVersion("0.1.0"),
		Capabilities: lifecycletypes.ServerCapabilities{
			Tools: cm.Some(lifecycletypes.ToolsCapability{
				ListChanged: cm.None[bool](),
			}),
			Experimental: cm.None[mcptypes.JSONObject](),
			Logging:      cm.None[mcptypes.JSONObject](),
			Completions:  cm.None[mcptypes.JSONObject](),
			Prompts:      cm.None[lifecycletypes.PromptsCapability](),
			Resources:    cm.None[lifecycletypes.ResourcesCapability](),
		},
		ServerInfo: lifecycletypes.Implementation{
			Name:       "weather-go",
			Version:    "0.1.0",
			Title:      cm.Some("Weather Go Provider"),
			WebsiteURL: cm.None[string](),
			Icons:      cm.None[cm.List[mcptypes.Icon]](),
		},
		Instructions: cm.Some("A Go MCP server providing weather tools"),
	}
	// In standard Go, this would be: return result, nil
	// The SetOK method sets the success variant of the Result type
	var res cm.Result[lifecycle.InitializeResultShape, lifecycletypes.InitializeResult, mcptypes.McpError]
	res.SetOK(result)
	return res
}

// ClientInitialized handles client initialization notification
func ClientInitialized() cm.Result[mcptypes.McpError, struct{}, mcptypes.McpError] {
	// Nothing to do on client initialization
	var result cm.Result[mcptypes.McpError, struct{}, mcptypes.McpError]
	result.SetOK(struct{}{})
	return result
}

// Shutdown handles shutdown request
func Shutdown() cm.Result[mcptypes.McpError, struct{}, mcptypes.McpError] {
	// Clean shutdown
	var result cm.Result[mcptypes.McpError, struct{}, mcptypes.McpError]
	result.SetOK(struct{}{})
	return result
}

File: templates/typescript/content/src/capabilities/lifecycle.ts
/**
 * Lifecycle capability implementation for {{project-name | kebab_case}} MCP server.
 * 
 * This module handles MCP session lifecycle: initialization, client ready, and shutdown.
 * In TypeScript/JavaScript with jco, we export functions that match the WIT interface
 * signatures. Unlike Python classes or Rust traits, jco uses plain function exports.
 */

import type {
  InitializeRequest,
  InitializeResult,
  ServerCapabilities,
  Implementation,
  ToolsCapability,
} from '../generated/interfaces/wasmcp-mcp-lifecycle-types.js';

/**
 * Initialize the MCP server.
 * 
 * TypeScript with jco handles Result types transparently - we return the success
 * value directly, and errors are thrown as exceptions. This is similar to Python's
 * componentize-py approach but different from Rust's explicit Result<T, E>.
 */
export function initialize(_request: InitializeRequest): InitializeResult {
  // Declare our capabilities - we support tools
  const toolsCapability: ToolsCapability = {
    listChanged: undefined,
  };

  const capabilities: ServerCapabilities = {
    experimental: undefined,
    logging: undefined,
    completions: undefined,
    prompts: undefined,
    resources: undefined,
    tools: toolsCapability,
  };

  const serverInfo: Implementation = {
    name: '{{project-name | kebab_case}}',
    version: '0.1.0',
    title: '{{project-name | kebab_case}} Provider',
    icons: undefined,
    websiteUrl: undefined,
  };

  return {
    protocolVersion: '0.1.0',
    capabilities,
    serverInfo,
    instructions: '{{project-description}}',
  };
}

/**
 * Called when the client has initialized.
 */
export function clientInitialized(): void {
  // No-op for this example
}

/**
 * Shutdown the server.
 */
export function shutdown(): void {
  // No-op for this example
}